### ISSUE TITLE
Smarter Receipt Validation

### ISSUE THREAD

**Issue by j6k4m8** [donatemates/donatemates#47]
Use the format of the receipt ID to validate which charityparser to use. For instance, ACLU receipt IDs are series of integers. PP uses ints and uppercase chars.

- [ ] receipt id
- [ ] date of receipt

**Comment by slavingia** [comment]
seems like the email address that receives the emails should determine this no? and the charity parser would fail if it doesn't match.

ie. if you create a charity for aclu, it'll fail if someone sends the corresponding email a pp receipt.

**Comment by dkleissa** [comment]
yeah, i feel like the receipt ID isn't a good way to pick the parser. Should be the other way around? doesn't seem like the receipt Id will be vary unique in many cases vs. lots of rich context in the email text and subject text.

**Comment by j6k4m8** [comment]
Mostly intended this as a double-check: If the receipt doesn't line up with
an expected format, then we shouldn't try running through parsers (which is
a computationally more complex task). So probably this lives in the
`is_receipt` fn.

Not high priority, but it's some validation that we should probably do
before we (eventually) hit an external resource with receipts to confirm

**Comment by dkleissa** [comment]
I think maybe this functionality fits in the class of "fraud" detectors which we should start to build out and run independently of detecting the charity and parsing the email.

**Comment by j6k4m8** [comment]
I like it — definitely a good home for this logic

**Comment by slavingia** [comment]
one validation that would be cool would be date validation:

the donation date must be AFTER the campaign creation date

**Comment by dkleissa** [comment]
other things to check are if the receipt id and date match in the HTML and plain-text versions of the email. Someone changing this in their email client typically would only edit one.

Think we should add a fraud detection method into the charity class that runs a list of methods that are unique to that charity. Also have a list of methods that run for all charities that could be in the parent class.

**Comment by craigjon** [comment]
Gentlemen - my apologies for intruding.  I read about Donatemates through Flipboard.  This seemed like the appropriate place to pass along the following.  The OrgHunter/Make My Donation platform deals with all of the above issues through APIs including:  selection of vetted 501(c)(3), recall of EIN for donation transactions, donation processing, funds disbursement to charities, receipting to donors, onboarding of first time donation recipients so the right people get the money, reporting to donors and charities, and compliance with all state and federal regulations (no small issue).  YES, it cost $$$ to integrate the platform into your app, BUT it is a one time fee, and thereafter if you choose you can receive a small percentage allowance per donation to offset your operating costs.  You can write me at canderson@makemydonation.org, if for no other reason than to say thanks but no thanks.

**Comment by slavingia** [comment]
We don't handle payments processing, we just work with receipts from the charities themselves. I don't think we'll go in that direction.

**Comment by craigjon** [comment]
Sahil,

Thanks for the quick response.

If you decide to go down that path, let me know.

Also, if you want to tinker with our APIs, feel free to register and play
in the sandbox.

Craig
